src/ApiCheck/BaselineGenerator.cs

```diff
@@ -0,0 +1,350 @@
+                    {
+                        var constraints = GetGenericConstraintsFor(method.GetGenericArguments().Select(t => t.GetTypeInfo()));
+                        foreach (var constraint in constraints)
+                        {
+                            methodBaseline.GenericConstraints.Add(constraint);
+                        }
+                    }
+
+                    methodBaseline.Static = method.IsStatic;
+                    methodBaseline.Sealed = method.IsFinal;
+                    methodBaseline.Virtual = method.IsVirtual;
+                    methodBaseline.Override = method.IsVirtual && method.GetBaseDefinition() != method;
+                    methodBaseline.Abstract = method.IsAbstract;
+                    methodBaseline.New = !method.IsAbstract && !method.IsVirtual && method.IsHideBySig &&
+                        method.DeclaringType.GetMember(method.Name).OfType<MethodInfo>()
+                        .Where(m => SameSignature(m, method)).Count() > 1;
+                    methodBaseline.Extension = method.IsDefined(typeof(ExtensionAttribute), false);
+
+                    foreach (var parameter in method.GetParameters())
+                    {
+                        var parameterBaseline = GenerateParameterBaseline(parameter);
+                        methodBaseline.Parameters.Add(parameterBaseline);
+                    }
+
+                    methodBaseline.ReturnType = TypeBaseline.GetTypeNameFor(method.ReturnType.GetTypeInfo());
+
+                    return methodBaseline;
+                case MemberTypes.Field:
+                    var field = (FieldInfo)member;
+                    var fieldBaseline = new MemberBaseline();
+
+                    fieldBaseline.Visibility = field.IsPublic ? BaselineVisibility.Public :
+                        field.IsFamilyOrAssembly ? BaselineVisibility.ProtectedInternal :
+                        field.IsFamily ? BaselineVisibility.Protected :
+                        field.IsPrivate ? BaselineVisibility.Private : BaselineVisibility.Internal;
+
+                    fieldBaseline.Constant = field.IsLiteral;
+                    fieldBaseline.Static = field.IsStatic;
+                    fieldBaseline.ReadOnly = field.IsInitOnly;
+                    fieldBaseline.Kind = MemberBaselineKind.Field;
+
+                    fieldBaseline.Name = field.Name;
+                    fieldBaseline.ReturnType = TypeBaseline.GetTypeNameFor(field.FieldType.GetTypeInfo());
+
+                    return fieldBaseline;
+                case MemberTypes.Event:
+                case MemberTypes.Property:
+                case MemberTypes.NestedType:
+                    // All these cases are covered by the methods they implicitly define on the class
+                    // (Properties and Events) and when we enumerate all the types in an assembly (Nested types).
+                    return null;
+                case MemberTypes.TypeInfo:
```

Comment:
That's good, but then wouldn't this code be a problem? This would presumably explode on nested types, no? I see there are some tests around this but I don't see how this code can work...